### ISSUE TITLE
STDLibLogObserver eats unhandled errors in Deferreds

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8037]
|<img alt="OptiverTimAll's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| OptiverTimAll reported|
|-|-|
|Trac ID|trac#8037|
|Type|defect|
|Created|2015-09-30 06:10:23Z|

Here's a simple test-case:

```
#!/usr/bin/env python2.7
import logging

from twisted import logger
from twisted.internet import defer


# Route Python's logging somewhere sensible.
logging.basicConfig()


# Route Twisted's logging to Python's logging system.
logger.globalLogBeginner.beginLoggingTo([logger.STDLibLogObserver()])


# Make a deferred with an unhandled error.
d = defer.Deferred()
d.addCallback(lambda _: does_not_exist())
d.callback(None)
```

With Twisted 15.4.0 on Python 2.7 (on CentOS 5, sigh), when I save the above script as ```ic-test.py``` and run it, I get the following unhelpful output:

```
CRITICAL:twisted:Unhandled error in Deferred:
CRITICAL:twisted:
```

If I comment out the line that calls ```beginLoggingTo()```, I get much more helpful output:

```
Unhandled error in Deferred:


Traceback (most recent call last):
  File "ic-test.py", line 18, in <module>
    d.callback(None)
  File "/home/timall/.virtualenvs/txfuse/lib/python2.7/site-packages/twisted/internet/defer.py", line 393, in callback
    self._startRunCallbacks(result)
  File "/home/timall/.virtualenvs/txfuse/lib/python2.7/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/home/timall/.virtualenvs/txfuse/lib/python2.7/site-packages/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "ic-test.py", line 17, in <lambda>
    d.addCallback(lambda _: does_not_exist())
exceptions.NameError: global name 'does_not_exist' is not defined
```

My [original mailing-list message](http://thread.gmane.org/gmane.comp.python.twisted/28486) suggested this was somehow an interaction between ```inlineCallbacks``` and the new logging system, but the behaviour is just the same with a bare Deferred.

<details><summary>Searchable metadata</summary>

```
trac-id__8037 8037
type__defect defect
reporter__OptiverTimAll OptiverTimAll
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__core core
keywords__None None
time__1443593423999302 1443593423999302
changetime__1503884458348890 1503884458348890
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

[#8264](https://github.com/twisted/twisted/issues/8264) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="jlitzingerdev's avatar" src="https://avatars.githubusercontent.com/u/1496650?s=50" width="50" height="50">](https://github.com/jlitzingerdev)<a name="note_2"></a>|@jlitzingerdev commented|
|-|-|

I'm wondering if this is fixed due to some change in python itself.

I get the helpful (not the Unandled error) message for this example as well as the example in [#8264](https://github.com/twisted/twisted/issues/8264) with:
* Current twisted trunk
* Twisted 15.4.0 tag
* python 2.7.13
* python 3.6.1

My host is Arch Linux, and the tests were run inside a virtualenv with the versions above.

I'll see if I can track down where this might have been fixed to understand how it relates to versions supported by Twisted.

Note:  I realize that doesn't necessarily mean the problem is fixed as older versions are still in use, but it does seem relevant to the ticket status.

**Comment by twisted-trac** [comment]
|[<img alt="jlitzingerdev's avatar" src="https://avatars.githubusercontent.com/u/1496650?s=50" width="50" height="50">](https://github.com/jlitzingerdev)<a name="note_3"></a>|@jlitzingerdev commented|
|-|-|

From what I've been able to reproduce, this error goes away with python 2.7.6 and still exists with 2.7.5.  This is true for both the current trunk as well as the reported (15.4.0) version (meaning, I think, the "fix" had nothing to do with Twisted changes).

My question now is whether this ticket should still be fixed, or whether 2.7.5 is unsupported?  Looking through one of the buildbot logs for RHEL7 [1] it looks (this assumes I read it correctly) like it is, which leads me to the question, is there interest in adding a fix to twisted that will work on 2.7.5 even though the issue will be resolved in 2.7.6?

I'm still going to look into what changed as that seems relevant for historical purposes even if no fix is added.

[1] https://buildbot.twistedmatrix.com/builders/rhel7-py2.7/builds/826/steps/select/logs/stdio

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Only the latest version of 2.7.x is supported at any given time, so if this really was just a bug in 2.7.5 and earlier, then this can be closed.

However, it might just be an accidental change in behavior of those versions, so let's be sure we know what the change was before we close this :).

**Comment by twisted-trac** [comment]
|[<img alt="jlitzingerdev's avatar" src="https://avatars.githubusercontent.com/u/1496650?s=50" width="50" height="50">](https://github.com/jlitzingerdev)<a name="note_5"></a>|@jlitzingerdev commented|
|-|-|

Thanks for the reply.  I'll track down the cause of the behavior change and update the ticket.

**Comment by twisted-trac** [comment]
|[<img alt="jlitzingerdev's avatar" src="https://avatars.githubusercontent.com/u/1496650?s=50" width="50" height="50">](https://github.com/jlitzingerdev)|@jlitzingerdev set owner to @jlitzingerdev|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jlitzingerdev's avatar" src="https://avatars.githubusercontent.com/u/1496650?s=50" width="50" height="50">](https://github.com/jlitzingerdev)<a name="note_7"></a>|@jlitzingerdev commented|
|-|-|

Short story - Pretty sure I was wrong, the fix is in twisted, not python itself.

Long story:

In digging in I discovered I had a bug in my setup.  Specifically, Twisted was installed in the virtualenv with the -e flag, but when I checked out 15.4.0 I did not re-execute the install.  Unfortunately, that resulted in the cases above "working" against a 15.4.0 tree (I'm guessing the bytecode from trunk was still there).

Still appears to be fixed, but due to a change in twisted itself.  Will bisect twisted to see where.

**Comment by twisted-trac** [comment]
|[<img alt="jlitzingerdev's avatar" src="https://avatars.githubusercontent.com/u/1496650?s=50" width="50" height="50">](https://github.com/jlitzingerdev)<a name="note_8"></a>|@jlitzingerdev removed owner|
|-|-|

Looks like it was pretty deliberately fixed to address [#8316](https://github.com/twisted/twisted/issues/8316).

https://github.com/twisted/twisted/commit/91755e8b141f50f17d9f3371238643c9b0a8a091

I wasn't sure whether it was valid for those without commit access to mark a ticket fixed, so I'm just putting this back into review.

**Comment by twisted-trac** [comment]
|[<img alt="twm's avatar" src="https://avatars.githubusercontent.com/u/43662?s=50" width="50" height="50">](https://github.com/twm)<a name="note_9"></a>|@twm set status to `closed`|
|-|-|

I think it's fine for non-committers to close bugs. Indeed, the help is welcome! In this case the correct course of action is to resolve as a duplicate of [#8316](https://github.com/twisted/twisted/issues/8316). For anyone who runs across this issue in the future: the fix looks to be in Twisted 16.2.0.

Thanks for using Twisted and for digging into this issue!
